File: backend/src/main/java/com/tripfriend/domain/review/repository/ReviewRepository.java
package com.tripfriend.domain.review.repository;

import com.tripfriend.domain.review.entity.Review;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface ReviewRepository extends JpaRepository<Review, Long> {

    // 댓글 수를 기준으로 리뷰 정렬 (많은 순)
    @Query("SELECT r, COUNT(c) as commentCount FROM Review r LEFT JOIN Comment c ON r.reviewId = c.review.reviewId GROUP BY r ORDER BY commentCount DESC")
    List<Object[]> findAllOrderByCommentCountDesc();

    // 평점 높은 순으로 정렬
    List<Review> findAllByOrderByRatingDesc();

    // 평점 낮은 순으로 정렬
    List<Review> findAllByOrderByRatingAsc();

    // 최신순 정렬 - 기본값
    List<Review> findAllByOrderByCreatedAtDesc();

    // 오래된순 정렬
    List<Review> findAllByOrderByCreatedAtAsc();

    // 제목으로 검색
    List<Review> findByTitleContainingOrderByCreatedAtDesc(String keyword);

    // 여행지 ID로 검색
    List<Review> findByPlace_IdOrderByCreatedAtDesc(Long placeId);

    // 여행지 ID + 정렬 옵션 (평점 높은 순)
    List<Review> findByPlace_IdOrderByRatingDesc(Long placeId);

    // 여행지 ID + 정렬 옵션 (평점 낮은 순)
    List<Review> findByPlace_IdOrderByRatingAsc(Long placeId);

    // 특정 사용자의 리뷰 목록
    List<Review> findByMemberIdOrderByCreatedAtDesc(Long memberId);
}

File: backend/src/main/java/com/tripfriend/domain/review/repository/ReviewRepository.kt
package com.tripfriend.domain.review.repository

import com.tripfriend.domain.review.entity.Review
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository

@Repository
interface ReviewRepository : JpaRepository<Review, Long> {

    // 특정 여행지 - 댓글 많은순 -> 최신순 정렬
    @Query("""
        SELECT r 
        FROM Review r 
        LEFT JOIN Comment c ON r.reviewId = c.review.reviewId 
        WHERE r.place.id = :placeId 
        GROUP BY r 
        ORDER BY COUNT(c) DESC, r.createdAt DESC
    """)
    fun findByPlace_IdOrderByCommentCountDesc(placeId: Long): List<Review>

    // 특정 여행지 - 평점 높은순 -> 조회수 많은순 -> 최신순
    @Query("""
        SELECT r 
        FROM Review r
        LEFT JOIN ReviewViewCount vc ON r.reviewId = vc.reviewId
        WHERE r.place.id = :placeId 
        ORDER BY r.rating DESC, vc.count DESC, r.createdAt DESC
    """)
    fun findByPlace_IdOrderByRatingDesc(placeId: Long): List<Review>

    // 특정 여행지 - 평점 낮은순 -> 조회수 많은순 -> 최신순
    @Query("""
        SELECT r 
        FROM Review r
        LEFT JOIN ReviewViewCount vc ON r.reviewId = vc.reviewId
        WHERE r.place.id = :placeId 
        ORDER BY r.rating ASC, vc.count DESC, r.createdAt DESC
    """)
    fun findByPlace_IdOrderByRatingAsc(placeId: Long): List<Review>

    // 특정 여행지 - 최신순
    fun findByPlace_IdOrderByCreatedAtDesc(placeId: Long): List<Review>

    // 특정 여행지 - 오래된순
    fun findByPlace_IdOrderByCreatedAtAsc(placeId: Long): List<Review>

    // 전체 - 평점 높은순 -> 조회수 많은순 -> 최신순
    @Query("""
        SELECT r 
        FROM Review r
        LEFT JOIN ReviewViewCount vc ON r.reviewId = vc.reviewId
        ORDER BY r.rating DESC, vc.count DESC, r.createdAt DESC
    """)
    fun findAllByOrderByRatingDesc(): List<Review>

    // 전체 - 평점 낮은순 -> 조회수 많은순 -> 최신순
    @Query("""
        SELECT r 
        FROM Review r
        LEFT JOIN ReviewViewCount vc ON r.reviewId = vc.reviewId
        ORDER BY r.rating ASC, vc.count DESC, r.createdAt DESC
    """)
    fun findAllByOrderByRatingAsc(): List<Review>

    // 전체 - 최신순
    fun findAllByOrderByCreatedAtDesc(): List<Review>

    // 전체 - 오래된순
    fun findAllByOrderByCreatedAtAsc(): List<Review>

    // keyword 포함된 리뷰를 최신순으로 정렬
    fun findByTitleContainingOrderByCreatedAtDesc(keyword: String): List<Review>

    // 특정 멤버의 리뷰 목록 최신순
    fun findByMemberIdOrderByCreatedAtDesc(memberId: Long): List<Review>

    // 전체 리뷰 댓글 많은순 → 최신순 정렬
    @Query("""
        SELECT r, COUNT(c) as commentCount 
        FROM Review r 
        LEFT JOIN Comment c ON r.reviewId = c.review.reviewId 
        GROUP BY r 
        ORDER BY commentCount DESC, r.createdAt DESC
    """)
    fun findAllOrderByCommentCountDesc(): List<Array<Any>>
}